lib/internal/test_runner/tap_checker.js

```diff
@@ -0,0 +1,128 @@
+'use strict';
+
+const { NumberParseInt } = primordials;
+const {
+  codes: { ERR_TAP_VALIDATOR_ERROR },
+} = require('internal/errors');
+
+class TAPValidationStrategy {
+  validate(ast) {
+    this.#validateDocument(ast);
+    this.#validateVersion(ast);
+    this.#validatePlan(ast);
+    this.#validateTestPoints(ast);
+
+    return true;
+  }
+
+  #validateDocument(ast) {
+    const { documents } = ast.root;
+
+    if (documents.length > 1) {
+      throw new ERR_TAP_VALIDATOR_ERROR('Found more than one TAP documents');
+    }
+  }
+
+  #validateVersion(ast) {
+    const { version } = ast.root.documents[0];
+
+    // TAP14 specification is compatible with observed behavior of existing TAP13 consumers and producers
+    if (version !== '14' && version !== '13') {
+      throw new ERR_TAP_VALIDATOR_ERROR('TAP version should be 14');
+    }
+  }
+
+  #validatePlan(ast) {
+    const { plan } = ast.root.documents[0];
+
+    if (!plan) {
+      throw new ERR_TAP_VALIDATOR_ERROR('Missing Plan');
+    }
+
+    if (!plan.start) {
+      throw new ERR_TAP_VALIDATOR_ERROR('Missing Plan start');
+    }
+
+    if (!plan.end) {
+      throw new ERR_TAP_VALIDATOR_ERROR('Missing Plan end');
+    }
+
+    const planStart = NumberParseInt(plan.start, 10);
+    const planEnd = NumberParseInt(plan.end, 10);
+
+    if (planEnd !== 0 && planStart > planEnd) {
+      throw new ERR_TAP_VALIDATOR_ERROR(
+        `Plan start ${planStart} is greater than Plan end ${planEnd}`
+      );
+    }
+  }
+
+  #validateTestPoints(ast) {
```

Comment:
@manekinekko wow :open_mouth: it works with `#`, looks nice, I didn't know about that, what the style rule for that, like public/private method diversity or so?